### DISCUSSION TITLE
Is it possible to avoid caching some symlinks with FUSE_CAP_CACHE_SYMLINKS?

### DISCUSSION THREAD

**Discussion by AM5800** [libfuse/libfuse#970] [Q&A]
Hello, 

we'd like to use FUSE_CAP_CACHE_SYMLINKS in our project.
To my understanding, with this cap, symlinks once read, can be potentially cached forever. But we have some symlinks that we'd like to not be cached at all(For example, because they contain the reference to the latest state and this moves too fast).

1) Is it possible to somehow not cache particular symlinks?
2) If not, what would be the right way to tackle this? Call `fuse_invalidate_path` after every `readlink`? Won't there be a race(i.e. one can't call `fuse_invalidate_path` from `readlink` itself, so we'd need to start a thread to do it. What if `fuse_invalidate_path` completes before the `readlink` itself?)

P.S. We use high-level libfuse interface

**Comment by bsbernd** [comment]
I think we need a different fuse mailing list on kernel.org, this is not a libfuse question, but kernel side. See `fuse_get_link`
```
static const char *fuse_get_link(struct dentry *dentry, struct inode *inode,
				 struct delayed_call *callback)
{
	struct fuse_conn *fc = get_fuse_conn(inode);
	struct page *page;
	int err;

	err = -EIO;
	if (fuse_is_bad(inode))
		goto out_err;

	if (fc->cache_symlinks)
		return page_get_link(dentry, inode, callback);

	err = -ECHILD;
	if (!dentry)
		goto out_err;

	page = alloc_page(GFP_KERNEL);
	err = -ENOMEM;
	if (!page)
		goto out_err;

	err = fuse_readlink_page(inode, page);
	if (err) {
		__free_page(page);
		goto out_err;
	}

	set_delayed_call(callback, page_put_link, page);

	return page_address(page);

out_err:
	return ERR_PTR(err);
}
```

We don't have a FUSE_DENTRY yet, so that would need to be set somehow into the inode (lookup flag?). Anyway this question should go to linux-fsdevel for now.

**Comment by ashleypittman** [comment]
You can't modify a symlink, you can only remove and replace it.  If you create a new symlink then it has a new inode number and that invalidates the cache.

To get the behaviour you want then you need to tell libfuse a different inode number for the symlink at the point where you want it to reference something else.

This is from the perspective of the kernel and the low level interface however, I'd need to check through the high-level code to see how this is supported there.

**Comment by bsbernd** [comment]
You would need to pass a flag on lookup and store it into the inode, but kernel side has zero support for that... As I said, this is kernel side question.

**Comment by bsbernd** [comment]
Regarding fuse_invalidate_path, right now with /dev/fuse it would work, if you would use the lowlevel interface, as you would then do fuse_reply_readlink() yourself, which is a /dev/fuse write operation. Once that has completed, you would need to do invalidate. With high level interface not possible, as libfuse itself calls fuse_reply_readlink() outside of your operation.
